### ISSUE TITLE
window.addEventListener('beforeunload', onbeforeunload) can't prevent the window from being closed

### ISSUE THREAD

**Issue by duanyao** [atom/electron#2481]
The following code can't prevent the window from being closed in Electron, although '>>>> onbeforeunload called' is shown.

```js
window.addEventListener('beforeunload', onbeforeunload);
function onbeforeunload(e) {
	console.log('>>>> onbeforeunload called');
	return false;
};
```

` window.onbeforeunload = onbeforeunload; ` works as expected however.

Tested with Electron 0.30.4 win32.

By the way, why `beforeunload` event works differently in Electron? If you absolutely can't make it spec-compliant, I suggest making a new Electron-specific event, and make `beforeunload` event a no-op.

**Comment by etiktin** [comment]
This happens also in v0.30.2 on Windows 8.1 64 bit.

I tried inspecting the `beforeunload` listeners using `getEventListeners(window)` and everything looks fine. Also tried using `Object.getOwnPropertyDescriptor(window, 'onbeforeunload')`, to see if there's a special setter for `onbeforeunload` but there isn't.

I agree that modifying the behavior of `beforeunload` is confusing (see #2361 for additional issues).

**Comment by deepak1556** [comment]
@duanyao you need to set the return value of event object in `beforeunload` handler to `false` or `empty` string inorder to prevent navigation.

```
window.addEventListener('beforeunload', onbeforeunload);
function onbeforeunload(e) {
    console.log('>>>> onbeforeunload called');
    e.returnValue = "false";
};
```

The spec [states](https://html.spec.whatwg.org/multipage/browsers.html#unloading-documents)

```
If the user did not confirm the page navigation, then the user agent refused to allow the document to be unloaded.
```

Since the developers are free to create their own confirmation dialogs in Electron, we require  the developers to return the result for the required action to proceed. Maybe this should be documented ?

**Comment by duanyao** [comment]
@deepak1556 
The spec says:

> If the returnValue attribute of the event object is not the empty string, or if the event was canceled, then the user agent should ask the user to confirm that they wish to unload the document.

So `e.returnValue = "false"` should result in a confirm dialog per spec; however Electron simply rejects the navigation/closing in this case.

So I suggest adding a Electron specific event, e.g. `electron-beforeunload` to allow custom dialogs, and make `beforeunload` spec-compaliant.

**Comment by deepak1556** [comment]
Proiding a new event doesnt sound good, if `event.returnValue` is something other than empty string or `false` Electron can fallback to provide its own dialog for confirmation. Does that sound good ? will wait for others thoughts.

**Comment by etiktin** [comment]
@deepak1556, in your example you wrote `e.returnValue = "false";`, didn't you mean `e.returnValue = false;`?

Why does returning `false` from `window.onbeforeunload` works, but we need to use `e.returnValue` from handlers added through `window.addEventListener`? I have no problem using `e.returnValue`, but I'm just curious why it works differently.

@duanyao, `BrowserWindow` already has a special event that allows you to prevent unload. See the [close](https://github.com/atom/electron/blob/master/docs/api/browser-window.md#event-close) event. That's why it's weird that we also mess with the normal behavior of `beforeunload`.

Regarding the following:
> providing a new event doesn't sound good, if event.returnValue is something other than empty string or false Electron can fallback to provide its own dialog for confirmation.

I would prefer that `beforeunload` will act exactly like it does in browsers. If we can't do that because it's a breaking change (although I think most developers will be for it), then yes I would go for what you're suggesting, but we need to make sure that the fallback confirmation dialog appears only once per propagation (see #2361 for details). 

I think it would be better if customization of dialogs (confirm navigation, alert, confirm and prompt), would be done by registering a handler. So for example if I want my alerts to have a custom title and icon, I could do something like this:
```javascript
mainWindow.onalert = function (msg) {
	var dialog = remote.require('dialog');
	dialog.showMessageBox(
		mainWindow,
		{
			type: 'none',
                        icon: './alert.png',
			title: 'Foobar',
			message: msg,
			buttons: ['Click me'],
			noLink: true
		}
	);
}
```

**Comment by duanyao** [comment]
@deepak1556 I'm not sure how fallback would work in your mind. Currently I have to do some UA-sniff:

```
function onBeforeUnload(ev) {
        var msg = '...';
	if (!config.isElectron) {       
		ev.returnValue = msg;
	} else {
		return confirm(msg);
	}
}
```

@etiktin It seems `close` event is unavailable to web content. `e.returnValue = false` is equivlant to `e.returnValue = "false"`, because `BeforeUnloadEvent.returnValue` is a string, and `false` is converted to string "false".

I agree that `beforeunload` should be web-compatible unless you have very good reason not to.

Customization of dialogs globally sounds a good idea. In fact gecko/xulrunner allow you do this, take the [geckofx code](https://bitbucket.org/geckofx/geckofx-33.0/src/3e2c6436b414984328d4e3cbee6dedf5f09a9e9d/Geckofx-Core/PromptService.cs) as an example.

**Comment by deepak1556** [comment]
@duanyao you shldnt need to UA sniff with the fallback, my suggestion was

```
ev.returnValue = '' // empty string run silently
ev.returnValue = 'false' // result from custom dialog to cancel
ev.returnValue = msg // create dailog with message for conformation
```
not favouring the false value, just dont want to break existing users using custom dialogs.


```
function onBeforeUnload(ev) {
    var msg = '...';   
     ev.returnValue = msg; // non-empty string create a dialog with message for confirmation
}
```

Having said that, a better approach would be to set custom dialogs via api as callback to `AtomJavaScriptDialogManager` as noted by @etiktin https://github.com/atom/electron/issues/2522 and use them or default to native, that way can be made web-compatible. If any of you would like a PR the required logic is [here](https://github.com/atom/electron/blob/master/atom/browser/atom_javascript_dialog_manager.cc#L25)

**Comment by duanyao** [comment]
@deepak1556 Thanks for explanation. I think it is acceptable to just reserve 'false' for Electron apps.

**Comment by zcbenz** [comment]
For the difference between `window.addEventListener` and `window.onbeforeunload`, it is not an Electron thing, it is actually designed by DOM and you can see the same behavior in every modern browser.

For the behavior of `beforeunload` handler, showing a confirmation dialog doesn't cover all use cases. Suppose you are writing a text editor, before you close or refresh the window you need to show a dialog to ask the users whether to save the file, and if they choose "Save" you also need to show another dialog to ask where to save the file, and finally do some filesystem operations, and it can not be solved by showing a confirmation dialog. And example is what Atom did:
https://github.com/atom/atom/blob/a13e0020796b323e3e6c85bfeb873dfb81e0a779/src/window-event-handler.coffee#L60

Allowing registering custom handler for the `beforeunload` dialog doesn't solve the problem completely, because the custom handler will run in the main process while all your data is in the renderer process. So can we make Electron run the custom handler in renderer process to solve it? No, it will just create a dead lock since the renderer process is already waiting for the main process to answer.

Making use of `unload` handler is not an options too, because there is an optimization in Chrome that will delay the call of `unload` handler after loading the new page.

For changing the behavior of return `false` in `beforeunload` handler, in normal browser returning `false` is the same with returning `"false"`, because the `beforeunload` handler converts all results to string type. In electron we treated `false` and `"false"` specially because many APIs just return boolean type and this makes things simpler. Is this a good behavior? Probably not. But it is too late to change that since lots of apps are already relying on it.

We can do breaking changes, but existing code must break very noisily, if we do a breaking change silently it would just result in lots of developers spending hours trying to find why their apps stop working.

**Comment by duanyao** [comment]
@zcbenz 
Wow, you are correct about the difference between `window.addEventListener('beforeunload')` and `window.onbeforeunload`, `OnBeforeUnloadEventHandler` has a special treatment in the spec:
https://html.spec.whatwg.org/#the-event-handler-processing-algorithm

However, for the behavior of `beforeunload` handler in Electron, I think the problem is that Electron already break existing web pages. If we assume that all codes served to Electron are designed with Electron in mind, then it is fine. But this is not the case. What if one want to build an app that may load arbitray web pages (My current project is one example)?

I think the reasonable trade-off has already been proposed by @Caged deepak1556, let me make it clearer: 

```JS
ev.returnValue = 'false' // do what Electron does currently

return 'false';  // do what Electron does currently

ev.returnValue = a_string_other_than_false; // do what the spec says

return a_string_other_than_false; // do what the spec says
```

What do you think?

**Comment by zcbenz** [comment]
@duanyao I think we should also handle the case of returning `true`, so for apps that returning booleans the behavior is not changed. Other parts of the proposal looks good to me, can you create a new issue for that?

**Comment by duanyao** [comment]
@zcbenz  Done with #2579.
